### ISSUE TITLE
Can distributed strategy solve VRAM runout isuue?

### ISSUE THREAD

**Issue by Dream7-Kim** [tensorflow/tensorflow#33322]
I have a complex function with tens of millions parameters and two hundreds of variables. I need to calculate the gradient of this function,but my Nvidia graphic card has not enough VRAM. I tried with two cards but it cannot solve the issue.
Two cards only accelerate the calculating speed, cannot solve memory problem, because every video card needs to save the same calculation graph, right?

Anyone can help me?

**Comment by gadagashwini** [comment]
@Dream7-Kim, Please provide the standalone code and also tensorflow version. Please go through this [link](https://www.tensorflow.org/guide/distributed_training). Thanks!

**Comment by Dream7-Kim** [comment]
@gadagashwini

```python
import numpy.random as rng

NUM_VAR = 10**5
NUM_ARG = 200

var = 10 * rng.rand(1, NUM_VAR)
arg = rng.rand(NUM_ARG, 10)
varf = open('variable.txt', 'w')
for vari in var[0]:
    varf.write(str(vari) + " ")
varf.close()

argf = open('argument.txt', 'w')
for i in range(10):
    # print(arg[:, i])
    for argi in arg[:, i]:
        argf.write(str(argi) + " ")
    argf.write("\n")
argf.close()
```

```python
import tensorflow as tf
import time
import numpy as np
import read_data
import write_res
import recordtime

tf.debugging.set_log_device_placement(True)
print("Num GPUs available: ", len(tf.config.experimental.list_physical_devices('GPU')))

var, arg = read_data.read_data()

def func(arg):
    divider = 0 # denominator
    numerator = 0
    temp = tf.matmul(arg, var)
    temp_sin = tf.math.sin(temp)
    temp_cos = tf.math.cos(temp)
    temp = tf.math.add(temp_cos, temp_sin)

    temp1 = tf.pow(temp, 2)
    divider = tf.reduce_sum(temp1)
    divider = tf.pow(divider, 1/2)

    temp2 = tf.reduce_sum(temp, 1)
    temp2 = tf.pow(temp2, 2)
    numerator = tf.reduce_sum(temp2)

    ress = tf.divide(numerator, divider)

    return tf.math.log(ress)

recordtime.record("-----------------------------tensorflow execution time--------------------------------")
print("-----------------------------tensorflow execution time--------------------------------")
for i in range(10):
    arg_tf = tf.Variable(initial_value=arg[:, i].reshape(200,1), dtype=tf.float64)
    strategy = tf.distribute.MirroredStrategy()
    
    start = time.time()
    with strategy.scope():
        with tf.GradientTape(persistent=True) as g:
            g.watch(arg_tf)
            fun = func(arg_tf)
        grad = g.gradient(fun, arg_tf)
    print(grad)
    end = time.time()
    recordtime.record("\t" + str(i+1) + "th execution time: " + str(end-start) + " s ... ... ...")
    print("\t" + str(i+1) + "th execution time: " + str(end-start) + " s ... ... ...")
    write_res.writeres(grad.numpy(), 'tfres.txt')
    del g
```

This is my TF version code to run. The first one is to generate random data to test.
I browsed the link you provided. But I cannot find appropriate solution for my problem. MirroredStrategey cannot solve this problem. It needs to use ```model parallelism``` . But support for ```model parallelism``` in TF is specialized for models of deep learning. And as I know it will be distribute layers of neural network to multiple devices, but I need to calculate the gradient of a function and its calculation graph will be very big. Have you any solution for this?
Thanks for you reply and help!

**Comment by guptapriya** [comment]
distribution strategy doesn't yet support model parallelism, so you're right that it cannot help with your use case currently. Have you looked at mesh TF (https://github.com/tensorflow/mesh)? It is a general model parallelism solution built on top of tensorflow.

**Comment by jvishnuvardhan** [comment]
Closing this out since I understand it to be resolved, but please let me know if I'm mistaken. Thanks!
